src/Extensions/Diagnostics.HealthChecks/Internal/HealthCheckPublisher.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Fabric;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.ObjectPool;
+using ServiceFabricHealth = System.Fabric.Health;
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks
+{
+	internal abstract class HealthCheckPublisher : IHealthCheckPublisher
+	{
+		protected readonly ObjectPool<StringBuilder> StringBuilderPool;
+
+		internal static string? HealthReportSourceId => nameof(HealthCheckPublisher);
+
+		internal const string HealthReportSummaryProperty = "HealthReportSummary";
+
+		public HealthCheckPublisher(ObjectPoolProvider objectPoolProvider) => StringBuilderPool = objectPoolProvider.CreateStringBuilderPool();
+
+		public abstract Task PublishAsync(HealthReport report, CancellationToken cancellationToken);
+
+		protected void PublishAllEntries(HealthReport report, Action<ServiceFabricHealth.HealthInformation> publishFunc,
```

Comment:
Having a protected abstract method might be better than `Action<ServiceFabricHealth.HealthInformation>`, but it up to you to decide what fit's better.

Reply:
Yeah, I considered that - however the problem is that publishFunc can be async (as in RestHealthPublisher) or not (as partition publisher) - so functional approach seems to solve problem with that difference

Reply:
`async` is not a part of function signature so you can define an abstract function like:
```
protected abstract Task ReportHelthInfoAsync(ServiceFabricHealth.HealthInformation information);
```
And then have two implementations, sync:
```
protected override Task ReportHelthInfoAsync(ServiceFabricHealth.HealthInformation information)
{
     // Sync code
     return Task.CompletedTask;
}
```
async:
```
protected override async Task ReportHelthInfoAsync(ServiceFabricHealth.HealthInformation information)
{
     // code with await
}
```

Reply:
Ok, it explains also why `Action` didn't have problem with two versions (with(out) `async`)

Reply:
I think that I would stay with that functional approach because `reportHealth` can change, so it would be passed each time.

Reply:
I'm not following your reasoning here, what do you mean by reasoning `reportHealth` can change, and how abstract method definition would be different from Activity here? (aside from the fact that Activity creates delegate object and closure each time).

Reply:
I mean that in order to create `reportHealth` from `			Action<SfHealthInformation> reportHealth = partition switch...` we need partition accessor. At this point when I'm looking at this I see possibility to create such override implementation but this need some reimplementation, I think. Should I do this or functional approach is fine?

Reply:
Okay, I'm fine with keeping it in Action for now, as long as we understand the downsides, we can improve it later.